low-level/sensors/impl/generic/node_sas_port.py

```diff
@@ -181,47 +184,141 @@ def initialize(self, conf_reader, msgQlist, product):
 
         return True
 
-    def check_and_send_alert(self, new_phy_link_count):
+    def sort_phy_list(self, phy_list):
+        """Sorts phy list based on phy num
+        """
+        # List to be returned
+        sorted_list = []
+        # Phy num to phy name mapping
+        phy_num_dict = {}
+        # List of phy nums
+        num_list = []
+        for phy in phy_list:
+            # Read phy num eg phy-0:12 -> 12
+            num = int(phy.split(':')[1])
+            num_list.append(num)
+            phy_num_dict[num] = phy
+        # Sort the numbers
+        num_list.sort()
+        for num in num_list:
+            sorted_list.append(phy_num_dict[num])
+        return sorted_list
+
+    def update_sas_ports_status(self):
+        """Reads current phy status and updates port connectivity status
+        """
+        # Assumption : phys will be present in multiples of 4
+        phy_list = []
+        for phy in self.phy_dir_to_linkrate_mapping.keys():
+            phy_list.append(phy)
+        phy_list = self.sort_phy_list(phy_list)
+        # Now we have a sorted list of phys
+        # Phys 0-3 for the 0th sas port, and so on in groups of 4 phys
+        self.num_sas_ports = len(phy_list)//4
+        # Counter tracking sas port number and status
+        i = 0
+        while i < self.num_sas_ports:
+            port_phy_list = phy_list[4*i:4*i+4]
```

Comment:
Too many data structures getting used here.

A simple list for HBA can be used, in which indexes can be referred as phy number & values can be status, represented by 0/1. Based on simple calculation further, if all phy numbers representing a port is seen down, so port can be marked down, or vice versa up or degraded for partial number of phys up.

> hba = [0,0,0,0, 1,1,1,1, 1,0,0,1, 1,1,1,1]


>>> hba = [0,0,0,0, 1,1,1,1, 1,0,0,1, 1,1,1,1]
>>>
>>> all(phy == 0 for phy in hba[0:3])
True
>>>
>>> all(phy == 1 for phy in hba[4:7])
True
>>>
>>> (1 in hba[8:11])
True
>>>